Bubble printing model

Halbaroth opened the pull request:
This PR solves the issue #833. The most of the refactoring have been done in OptimAE but still AE prints model in the library. Now, it will make this decision in the binary only.

Notice that AE printed model after getting timeout. I didn't keep this behaviour but I can restore it if you prefer. In my opinion, if we get a model after a timeout, the model is probably wrong and shouldn't be outputted.

Halbaroth committed:
Bubbling the model printing in the binary

This commit bubbles the decision of printing models in the binary. Now,
the library do not print models at all.

Halbaroth committed:
Update the JS worker

We have to update the JS worker to dump models as we don't print models
in the library anymore.

Halbaroth commented:
Oups, I realize that I don't print timeout reason anymore as it's done during printing model ...

Halbaroth committed:
Save a timeout reason in the exception Util.Timeout

In some situations, we need to retrieve the reason of a timeout in the
binary but we have no environment of the SAT solver in these locations.

I add an argument to the timeout exception to save its reason.

Halbaroth committed:
Revert "Save a timeout reason in the exception Util.Timeout"

This reverts commit dbfed1b65a7636a359e7aee22d03713018529070.

Halbaroth commented:
In fact, we don't need to print the timeout reason in the binary when we catch the exception `Util.Timeout`. Indeed, in this case it could mean only one thing: Alt-Ergo received the alarm signal and the reason is `Signal`. Thus I revert my last commit. If we raise another timeout in the SAT solver, we convert it to `Unknown env` in `Frontend` and the reason will be print in `solving_loop`.

Stevendeo commented on src/lib/reasoners/sat_solver_sig.ml:
```diff
@@ -35,12 +35,22 @@ type timeout_reason =
 type unknown_reason =
   | Incomplete
   | Memout
   | Timeout of timeout_reason
```
Why not `[@@deriving show]` ?

Halbaroth replied:
Not sure to understand your remark :D

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -153,6 +153,23 @@ let main () =
+  let print_model ppf (Model ((module SAT), env)) =
+    let ur = SAT.get_unknown_reason env in
+    Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+      "@[<v 0>; Returned unknown reason = %a@]"
```
There shouldn't be `;` comments on the diagnostic channel.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -153,6 +153,23 @@ let main () =
+    match SAT.get_model env with
+    | None ->
+      Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+        "@[<v 0>It seems that no model has been computed so \
```
This box is not closed.

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -837,21 +857,14 @@ let main () =
-          | None ->
-            (* TODO: add the location of the statement. *)
-            recoverable_error "No model produced."; st
+          let _ = match State.get partial_model_key st with
+            | Some model ->
+              print_model (Options.Output.get_fmt_regular ()) model
+            | None ->
+              (* TODO: add the location of the statement. *)
+              recoverable_error "No model produced."
+          in
+          st
```
Not convinced we should print the unknown reason when we have a model, it will just be noise. If the user wants to see the unkonwn reason, they can call `(get-info :unknown-reason)`, we shouldn't force it down their throat.

(For `--dump-models` I am not entirely convinced either but it kind of makes sense b/c the user has no control & it helps distinguish the intermediate models from the final models with `--interpretation every`, but for `(get-model)` there is a mechanism in place to display that information with `(get-info :unknown-reason)` already)

bclement-ocp commented on src/bin/common/solving_loop.ml:
```diff
@@ -153,6 +153,23 @@ let main () =
+      Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+        "@[<v 0>It seems that no model has been computed so \
+         far. You may need to change your model generation strategy \
+         or to increase your timeouts."
```
Why change the error message?

Halbaroth replied:
I didn't change it but I print the timeout reason just before. Now, if we don't print always the `unknown reason` while outputting the model, it makes sense to keep print it if we fail to produce a model here.

bclement-ocp reviewed (changes requested):
Not a very big fan of duplicating the `print_model` function, could we keep it in the frontend and call it from both binary & worker?

Halbaroth committed:
Review changes

bclement-ocp commented on src/bin/js/worker_js.ml:
```diff
@@ -120,6 +120,22 @@ let main worker_id content =
       Frontend.print_status status n
     in
 
+    let print_model ppf env =
+      let ur = SAT.get_unknown_reason env in
+      Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+        "@[<v 0>; Returned unknown reason = %a@]"
+        Sat_solver_sig.pp_unknown_reason_opt ur;
+      match SAT.get_model env with
+      | None ->
+        Printer.print_fmt (Options.Output.get_fmt_diagnostic ())
+          "@[<v 0>It seems that no model has been computed so \
+           far. You may need to change your model generation strategy \
+           or to increase your timeouts."
+
+      | Some (lazy model) ->
+        Models.output_concrete_model ppf model
+    in
+
```
All the remarks made in the previous round on the version of this function in `solving_loop.ml` apply. Please either make the same changes as in `solving_loop.ml` or (better, because it would avoid this type of issues) keep the implementation in the library and call it from both `solving_loop.ml` and `worker.ml`.

Halbaroth replied:
I know ;) actually it's a bit annoying to factorize this function because of the presence of the first-class module as argument. I've several options but none of them are really satisfying. If I move back the function `print_model` in the module `Frontend`, I will need to recreate the module `Frontend` in the `get-model` handler as this function requires the module `SAT`. 
```ocaml 
          let () = match State.get partial_model_key st with
            | Some (Model ((module SAT), env)) ->
              let module FE = Frontend.Make (SAT) in
              FE.print_model (Options.Output.get_fmt_regular ()) env
            | None ->
              (* TODO: add the location of the statement. *)
              recoverable_error "No model produced."
          in
          st
```
I don't know what's the overhead of recreate a module here.

I can also retrieve the model in the `Solving_loop` and feed the function `print_model` with its return but `print_model` is just a wrapper around the pretty print of models to display a warning if we don't produce a model as we should according to the option `produce-models` and the status of the last `check-sat`. Thus, we have to feed the function `print_model` with an `option`, which is a bit ugly.

Halbaroth replied:
I factorize the code using the first method I described.

bclement-ocp replied:
There is no references in the frontend so it would be fine to re-create a `Frontend.Make` here, and the cost would be virtually zero compared to actually computing the model.

Having a `print_model_opt` in `Frontend` (not `Frontend.Make`) would also work. `Option.map` is not ugly :)

bclement-ocp replied:
Actually `print_model` in `Frontend` could take the first-class module, which would probably be the best solution.

Halbaroth replied:
I thought I could do that but `Frontend` is already a functor taking the `SAT` module as argument. Besides, if I want to do that, I will need to move the gadt you created in `Solving_loop` into `Frontend` and expose it here. I prefer using the function `Frontend.Make`. If the cost is negligible, it's acceptable.

Halbaroth committed:
Factorizing `print_model`

We use `print_model` twice (in the usual binary and the JS worker), thus
we can factorize it by moving it in `Frontend`.
